### PR TITLE
enhancement/12268-stock-tools-multiple-axes

### PR DESCRIPTION
**Pull request by raf18seb**
Improved #12268, added possibility to draw Stock Tools on multiple axes.

~~Initially, this is a draft PR. The commit has been made with `--no-verify` flag because of some TS errors - it is being investigated now.~~

~~Potential code review ideas will be discussed later.~~

~~FYI @goransle and @bre1470 - this is the branch we were discussing my TS issue on.~~

### CODE REVIEW

**Comment by bre1470** [comment]
The test are passing for me on this branch. Maybe ts-node behaves different on a Windows environment. 🤔
Does `npx gulp test` passes for you?

**Comment by raf18seb** [comment]
> Does npx gulp test passes for you?

No, it throwns another error:
```
TypeError: test\karma-conf.js: Emit skipped
    at getOutput (C:\Users\Rafal\Desktop\repos\highcharts\node_modules\ts-node\src\index.ts:755:17)
    at Object.compile (C:\Users\Rafal\Desktop\repos\highcharts\node_modules\ts-node\src\index.ts:968:32)
```

Goran is going to take a look at it on Windows.

**Comment by raf18seb** [comment]
The `getAssignedAxis` method checks whether (x, y) coordinates are within the axes `min` and `max` extremes. I am opened to discuss other approaches how we can valid coordinates.

File: ts/Extensions/Annotations/NavigationBindings.ts
Comment (karolkolodziej) [inline ts/Extensions/Annotations/NavigationBindings.ts:1270]: I would add the same comment everywhere you have that `if`. Not everyone is looking through the whole file, very often we concentrate on one method and I think we should have similar comments.

**Review by karolkolodziej — CHANGES_REQUESTED** [review]

File: ts/Extensions/Annotations/Annotations.ts
Comment (raf18seb) [reply ts/Extensions/Annotations/Annotations.ts:182]: It's index. I didn't change the name because it was named like this previously. We can consider changing it, but there are so many places when an axis' index is called this way that I think it should be a part of potential Stock Tools refactoring in the future.

Comment (pawelfus) [reply ts/Extensions/Annotations/Annotations.ts:182]: It's the same as with option `series.xAxis` or `series.yAxis`. It's an index, even when the name does not suggest it.

However, do we have `shapeOptions.xAxis` or `shapes[index].xAxis`? Only `shapes[index].point` or `points` have this option, according to the API: https://api.highcharts.com/highstock/annotations.shapes.point.xAxis (commented with `(1)` the same places)

And the format is `number | string | null`, could you double-check this?

File: ts/Extensions/Annotations/NavigationBindings.ts
Comment (pawelfus) [inline ts/Extensions/Annotations/NavigationBindings.ts:1283]: See my comment above, could you double-check this? I don't recall this option

File: ts/Extensions/Annotations/NavigationBindings.ts
Comment (pawelfus) [inline ts/Extensions/Annotations/NavigationBindings.ts:199]: Any particular reason to use `NavigationBindings.prototype.getAssignedAxis` over `chart.navigationBindings.getAssignedAxis` in this one annotation?

I'm fine with static methods, but then assign it to `NavigationBindings` instead of `NavigationBindings.prototype`.

File: ts/Stock/StockToolsBindings.ts
Comment (pawelfus) [inline ts/Stock/StockToolsBindings.ts:420]: `NavigationBindings.prototype.getAssignedAxis` vs `chart.navigationBindings.getAssignedAxis`

**Review by pawelfus — COMMENTED** [review]
Amazing! 🎉 
More questions than changes, see below

File: ts/Stock/StockToolsBindings.ts
Comment (pawellysy) [inline ts/Stock/StockToolsBindings.ts:789]: is it necessary to declare the options here, instead of after the if statement, where the options var is assigned?
Additionally, why not `const`?

**Review by pawellysy — APPROVED** [review]
Overall good, one question regarding the declaration of the variable.

File: ts/Extensions/Annotations/Annotations.ts
Comment (TorsteinHonsi) [reply ts/Extensions/Annotations/Annotations.ts:182]: > And the format is number | string | null, could you double-check this?

When `series.xAxis` or `series.yAxis` is a string, it refers to the `id` of the axis. For consistency, I think we should recreate that feature for the Annotations too. 

Also, I think we can simplify a bit and disallow `null`. I think 0 has the same effect, right?

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Please resolve comments and merge in master

File: ts/Extensions/Annotations/Annotations.ts
Comment (raf18seb) [reply ts/Extensions/Annotations/Annotations.ts:182]: Nice catch, @pawelfus 
I corrected TS definitions and now I'm taking axis index from shapes[index] point. Also, the API description should be corrected or even removed (the user doesn't need to define https://api.highcharts.com/highstock/annotations.shapes.point.xAxis because the click finds a proper axis automatically.

Comment (raf18seb) [reply ts/Extensions/Annotations/Annotations.ts:182]: Nice catch, @pawelfus
I corrected TS definitions and now I'm taking axis index from shapes[index] point. Also, the API description should be corrected or even removed (the user doesn't need to define https://api.highcharts.com/highstock/annotations.shapes.point.xAxis because the click finds a proper axis automatically.

Comment (raf18seb) [reply ts/Extensions/Annotations/Annotations.ts]: I think this option can be removed from API anyway because now the user doesn't need to define on which axis he is adding an annotation - new functionality finds the proper axis automatically.

When it comes to TS definitions, I think the Annotation's MockPoint's `xAxis` or `yAxis` props are only `number | AxisType` type now. I removed `string` and `null` from TS definitions.

Are you OK with these changes? https://github.com/highcharts/highcharts/pull/15209/commits/fd88d5ef63d43b515446d12f904475c0faed7b23

File: ts/Stock/StockToolsBindings.ts
Comment (raf18seb) [reply ts/Stock/StockToolsBindings.ts:789]: It's a habit related to `var` being hoisted. I didn't change `var` to `let` or `const` because this ticket is not about refactoring. If I wrote it from scratch, I would definitively write it your way.

**Comment by raf18seb** [comment]
Do you know where those CI errors come from? Something wrong with merge?

**Comment by bre1470** [comment]
The annotation type tests break:
```
test/typescript/modules/annotations.ts(37,22): error TS2769: No overload matches this call.
  Overload 1 of 2, '(options: Options, callback?: ChartCallbackFunction | undefined): Chart', gave the following error.
    Type '"container"' has no properties in common with type 'Options'.
```
Run `npx gulp lint-dts` to test local.

Also visual tests fail:
```
HeadlessChrome 89.0.4389 (Linux 0.0.0) Highcharts **********/annotations-advanced/measure FAILED
	Died on test #1     at samples/**********/annotations-advanced/measure/demo.js.preprocessed.js:2:15: Cannot read property 'toPixels' of undefined
	TypeError: Cannot read property 'toPixels' of undefined
	    at Measure.addCrosshairs (code/modules/annotations-advanced.src.js:5347:38)
	    at Measure.addShapes (code/modules/annotations-advanced.src.js:5319:22)
	    at Measure.Annotation.init (code/modules/annotations-advanced.src.js:2593:22)
	    at Measure.init (code/modules/annotations-advanced.src.js:5193:43)
	    at Measure.Annotation (code/modules/annotations-advanced.src.js:2582:22)
	    at new Measure (code/modules/annotations-advanced.src.js:5181:31)
	    at Chart.initAnnotation (code/modules/annotations-advanced.src.js:3550:34)
	    at code/modules/annotations-advanced.src.js:3630:44
	    at Array.forEach (<anonymous>)
	    at Chart.<anonymous> (code/modules/annotations-advanced.src.js:3624:39)
	HeadlessChrome 89.0.4389 (Linux 0.0.0): Executed 57 of 828 (1 FAILED) (skipped 8) (0 secs / -836235 mins 52.162 secs)
HeadlessChrome 89.0.4389 (Linux 0.0.0) Highcharts **********/annotations-advanced/pitchfork FAILED
	Died on test #1     at samples/**********/annotations-advanced/pitchfork/demo.js.preprocessed.js:2:15: Cannot read property 'left' of null
	TypeError: Cannot read property 'left' of null
	    at Function.InfinityLine.findEdgePoint (code/modules/annotations-advanced.src.js:4501:38)
	    at Pitchfork.middleLineEdgePoint (code/modules/annotations-advanced.src.js:4850:37)
	    at ControllablePath.point (code/modules/annotations-advanced.src.js:1040:56)
	    at ControllablePath.linkPoints (code/modules/annotations-advanced.src.js:1061:34)
	    at ControllablePath.init (code/modules/annotations-advanced.src.js:940:22)
	    at new ControllablePath (code/modules/annotations-advanced.src.js:1519:22)
	    at Pitchfork.Annotation.initShape (code/modules/annotations-advanced.src.js:2846:29)
	    at Pitchfork.addLines (code/modules/annotations-advanced.src.js:4871:22)
	    at Pitchfork.addShapes (code/modules/annotations-advanced.src.js:4867:22)
	    at Pitchfork.Annotation.init (code/modules/annotations-advanced.src.js:2593:22)
```

File: ts/Extensions/Annotations/Annotations.ts
Comment (pawelfus) [inline ts/Extensions/Annotations/Annotations.ts:1353]: `labels[index].point.xAxis` is supported: https://api.highcharts.com/highstock/annotations.labels.point.xAxis

File: ts/Extensions/Annotations/Types/Measure.ts
Comment (pawelfus) [inline ts/Extensions/Annotations/Types/Measure.ts:836]: `measure.typeOptions.xAxis` is supported: https://api.highcharts.com/highstock/annotations.measure.typeOptions.xAxis

File: ts/Extensions/Annotations/Types/CrookedLine.ts
Comment (pawelfus) [inline ts/Extensions/Annotations/Types/CrookedLine.ts:171]: I think `typeOptions` does support `xAxis` and `yAxis`: https://api.highcharts.com/highstock/annotations.crookedLine.typeOptions.xAxis

File: ts/Extensions/Annotations/NavigationBindings.ts
Comment (pawelfus) [inline ts/Extensions/Annotations/NavigationBindings.ts:995]: Can we move this function to `bindingUtils`?

**Review by pawelfus — CHANGES_REQUESTED** [review]
Just a minor polishing

File: ts/Extensions/Annotations/Types/Measure.ts
Comment (raf18seb) [reply ts/Extensions/Annotations/Types/Measure.ts:836]: Right, totally forgot that annotations still can be drawn in the chart's config...

File: samples/unit-tests/stock-tools/bindings/demo.js
Comment (pawellysy) [inline samples/unit-tests/stock-tools/bindings/demo.js:438]: Why you split declaration and assigning of the values to those variables? Can't you do assign them right away?

File: ts/Extensions/Annotations/NavigationBindings.ts
Comment (pawellysy) [inline ts/Extensions/Annotations/NavigationBindings.ts:257]: I find the `xOrY` parameter name a bit odd. Couldn't it be something like `isXAxis: boolean` like in Axis class?

**Review by pawellysy — CHANGES_REQUESTED** [review]
see comments, apart from that general question: Why are you splitting declaration and assignment of the variables? in some examples it just makes additional variable, which in case of early return statement call is redundant.

File: ts/Extensions/Annotations/NavigationBindings.ts
Comment (raf18seb) [reply ts/Extensions/Annotations/NavigationBindings.ts:257]: `axis.isXAxis` would not be useful in this case, because I am not operating on a specific axis, but on an object like this:
```js
coords = {
    xAxis: [{axis: axis, value: number}, ...],
    yAxis: [{axis: axis, value: number}, ...]
}
```

However, when I'm looking at this code after few days, I see that I don't need to loop through both properties `xAxis` and `yAxis` in `objectEach` there, because I am returning only one of them and I know which one from the beginning. I'll take a closer look at it tomorrow and probably improve.

**Comment by raf18seb** [comment]
@TorsteinHonsi please hold merging, I'll improve something here.

**Review by pawelfus — APPROVED** [review]
Approved!

**Comment by raf18seb** [comment]
@TorsteinHonsi ready to merge if no other changes requests.

**Review by karolkolodziej — APPROVED** [review]

**Review by pawellysy — APPROVED** [review]
Looks Good! Nothing to add :)

File: ts/Extensions/Annotations/MockPoint.ts
Comment (TorsteinHonsi) [inline ts/Extensions/Annotations/MockPoint.ts:233]: I don't think we should do the cast here... We don't _know_ that the series has an axis. What happens if the series is a pie series? Instead of casting, we should have a type check or allow the `AnnottationMockPointOptionsObject.xAxis` and `yAxis` to be undefined.

File: ts/Extensions/Annotations/NavigationBindings.ts
Comment (TorsteinHonsi) [inline ts/Extensions/Annotations/NavigationBindings.ts:257]: In this case, as the previous comment, we cannot cast `axis.min` and `axis.max` to `number`, because they aren't necessarily numbers. Instead we need to code for the case that they are not numbers.

In practice, we can test this by setting `data: []` in the given unit test.

File: ts/Extensions/Annotations/NavigationBindings.ts
Comment (TorsteinHonsi) [inline ts/Extensions/Annotations/NavigationBindings.ts:1300]: As above, avoid casting and instead do type checking.

File: ts/Extensions/Annotations/NavigationBindings.ts
Comment (TorsteinHonsi) [inline ts/Extensions/Annotations/NavigationBindings.ts:1379]: This block can be shortened considerably.

```ts
const xAxis = coordsX.axis.options.index;
const yAxis = coordsY.axis.options.index;
...
points: [
    { xAxis, yAxis, x, y },
    { xAxis, yAxis, x, y },
...
```

File: ts/Extensions/Annotations/Types/Pitchfork.ts
Comment (TorsteinHonsi) [inline ts/Extensions/Annotations/Types/Pitchfork.ts:129]: As above, avoid casting and instead do type checking.

File: ts/Extensions/Annotations/Types/VerticalLine.ts
Comment (TorsteinHonsi) [inline ts/Extensions/Annotations/Types/VerticalLine.ts:73]: As above, avoid casting and instead do type checking.

File: ts/Extensions/Annotations/Types/VerticalLine.ts
Comment (TorsteinHonsi) [inline ts/Extensions/Annotations/Types/VerticalLine.ts:98]: As above, avoid casting and instead do type checking.

File: ts/Stock/StockToolsBindings.ts
Comment (TorsteinHonsi) [inline ts/Stock/StockToolsBindings.ts:419]: Better do one check for `chart.navigationBindings` then assign coordsX and coordsY. Optional chaining is quite bulky when transpiled to ES5, and here it's twice over.

File: ts/Stock/StockToolsBindings.ts
Comment (TorsteinHonsi) [inline ts/Stock/StockToolsBindings.ts:449]: I know this is not new to this PR, but what if `closestPoint` is undefined? That case should probably be handled. Imagine that the series has `data: []`.

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks, great work! I added a bunch of comments, but they are all basically the same request: To avoid treating types of `(SomeType|undefined)` like `SomeType`, but also code for the case that it can be undefined.

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks for the updates! Please resolve conflicts before we merge.

**Review by TorsteinHonsi — APPROVED** [review]

**Pull request merged**
